File: python/bundle_spec/taxonomy.py
# SPDX-License-Identifier: MIT
# Copyright (c) 2026 Simon Morley / NullRabbit
"""Family taxonomy for bundle primitive classification.

The format targets the systems-software layer of validator
infrastructure. The same vulnerability classes you'd find in any
networked OSS daemon — response amplification, compute amplification,
memory amplification, connection exhaustion — just applied to
validator software specifically.

``family_id`` is the chain-agnostic vulnerability-class label.
``primitive_id`` (field on the manifest) is the chain-specific
implementation of a family (e.g. ``sui_F10_multi_get_objects_amp``
under ``family_id=response_amp``).

Models can train against either: family-level classification asks
"what kind of attack is this?"; primitive-level asks "which specific
known exploit is this?". Cross-chain holdout validates the format by
testing whether a model trained on Sui primitives under a family
generalises to a Solana primitive under the same family.

The enumeration is tight on purpose — each family is backed by at
least one real primitive, OR represents a pre-defined threat-model
slot whose population is intentionally deferred. Empty enum members
would be a contract claim that can't be validated, and the published
taxonomy reads as a research-scope claim. Tight beats aspirational.
Add a new family only when a primitive exists that doesn't fit any
current family, OR when the family represents a pre-defined
threat-model slot whose population is intentionally deferred.

``family_id`` is the **validator / node-software** vulnerability-class
vocabulary. On-chain contract / protocol / DeFi-economic findings (AMM
value-extraction, oracle / TWAP manipulation, liquidation, governance
capture, bridge-*contract* message abuse, ERC-4626 / reentrancy logic)
are **out of scope** — a separate DeFi-economic namespace, never
``family_id`` values. The VII ``ledger/family_map.csv`` is the
authoritative cross-map.

If a family name changes, bump ``BUNDLE_VERSION`` — it's a breaking
contract change for every bundle in the corpus.
"""

from __future__ import annotations

from enum import Enum


class FamilyId(str, Enum):
    """Chain-agnostic vulnerability-class label for a bundle.

    Enumeration: 10 attack families + 1 benign. ``reconnaissance`` is
    a deferred-population slot in v0.1.0 — the scaffolded family for
    pre-discovery and target-mapping activity (covered in upstream
    research but currently unpopulated in the public reference
    corpus).
    """

    response_amp = "response_amp"
    compute_amp = "compute_amp"
    memory_amp = "memory_amp"
    connection_exhaustion = "connection_exhaustion"
    consensus_abuse = "consensus_abuse"
    gossip_abuse = "gossip_abuse"
    auth_bypass = "auth_bypass"
    rate_limiter_bypass = "rate_limiter_bypass"
    service_misconfig = "service_misconfig"
    reconnaissance = "reconnaissance"
    benign = "benign"


FAMILY_DEFINITIONS: dict[FamilyId, str] = {
    FamilyId.response_amp:
        "Small request → disproportionately large response bytes. "
        "Attacker-side upstream cost is trivial; target-side egress "
        "dominates.",
    FamilyId.compute_amp:
        "Small request → disproportionate server-side CPU. Input size "
        "is small, processing cost is large (filter misses, parser "
        "pathologies, verification fallbacks, sync-over-async wedges).",
    FamilyId.memory_amp:
        "Unbounded accumulation of state in process memory. The "
        "attacker's input does not need to shrink, but every subsequent "
        "observation adds to retained state without bound.",
    FamilyId.connection_exhaustion:
        "Per-connection or per-subscription resource leak. Opening or "
        "subscribing consumes quota that isn't released on disconnect.",
    FamilyId.consensus_abuse:
        "Validator-authenticated misbehaviour inside the consensus "
        "protocol — equivocations, backpressure injection, proposal "
        "poisoning. Presupposes validator key material.",
    FamilyId.gossip_abuse:
        "Unauthenticated p2p / gossip protocol abuse — state-sync "
        "floods, request amplification via public peer endpoints.",
    FamilyId.auth_bypass:
        "Missing or broken authentication on a control, admin, or "
        "first-party API surface that should not be reachable without "
        "credentials.",
    FamilyId.rate_limiter_bypass:
        "Defect in rate-limiting logic allowing amplified or higher-"
        "than-intended traffic through the limiter — fail-open under "
        "errors, burst-doubling in token-bucket math, key-collision "
        "that shares quota across principals. Distinct from "
        "response_amp: the mechanism is a rate-limiter bug, the "
        "detection feature is limiter-state desync, and the mitigation "
        "is in the limiter code path — not a response-size cap.",
    FamilyId.service_misconfig:
        "Exploitable misconfiguration of an ancillary daemon that "
        "validators commonly co-locate (Redis, Elasticsearch, Grafana, "
        "Prometheus, SSH, …). Distinct from `reconnaissance` — "
        "`service_misconfig` is the exploitation of a discovered "
        "misconfig; `reconnaissance` is the prior-stage discovery "
        "traffic itself.",
    FamilyId.reconnaissance:
        "Pre-discovery and target-mapping activity against validator "
        "infrastructure. Spans the full spectrum from conventional "
        "active scanning (port enumeration, service fingerprinting, "
        "JSON-RPC method discovery) to the stealth end of the "
        "spectrum: low-frequency jittered probing, distributed low-"
        "volume reconnaissance, and timing-resonance fingerprinting "
        "that operates below the detection thresholds of conventional "
        "perimeter tooling. Included in the v1.0 taxonomy because "
        "off-the-shelf detection at the stealth end is unreliable in "
        "production deployments, and validator infrastructure is "
        "exactly the threat profile where patient adversarial pre-"
        "positioning matters. Distinct from `service_misconfig` "
        "(which is exploitation of a discovered weakness, not the "
        "discovery itself). Currently unpopulated in the v0.1.0 "
        "reference corpus.",
    FamilyId.benign:
        "No attack shape. Present to train the 'not malicious' decision "
        "boundary.",
}


assert set(FAMILY_DEFINITIONS.keys()) == set(FamilyId), (
    "every FamilyId must have a definition in FAMILY_DEFINITIONS"
)

File: rust/bundle_spec/src/taxonomy.rs
// SPDX-License-Identifier: MIT
//! Family taxonomy — chain-agnostic vulnerability-class label.
//!
//! Mirrors `python/bundle_spec/taxonomy.py`. Cross-language enum
//! consistency is pinned by the JSON Schema contract: both
//! languages emit the same wire string for each variant.

use serde::{Deserialize, Serialize};

/// Chain-agnostic vulnerability-class label for a bundle.
///
/// Enumeration: 10 attack families + 1 benign. `reconnaissance` is a
/// deferred-population slot in v0.1.0 — the scaffolded family for
/// pre-discovery and target-mapping activity.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum FamilyId {
    /// Small request → disproportionately large response bytes.
    ResponseAmp,
    /// Small request → disproportionate server-side CPU.
    ComputeAmp,
    /// Unbounded accumulation of state in process memory.
    MemoryAmp,
    /// Per-connection or per-subscription resource leak.
    ConnectionExhaustion,
    /// Validator-authenticated misbehaviour inside consensus.
    ConsensusAbuse,
    /// Unauthenticated p2p / gossip protocol abuse.
    GossipAbuse,
    /// Missing or broken authentication on a control surface.
    AuthBypass,
    /// Defect in rate-limiting logic allowing amplified traffic.
    RateLimiterBypass,
    /// Exploitable misconfiguration of an ancillary daemon.
    ServiceMisconfig,
    /// Pre-discovery / target-mapping activity.
    Reconnaissance,
    /// Subscription/streaming filter with disproportionate per-event CPU.
    SubscriptionCpuAmp,
    /// Abuse of the state / snapshot bootstrap-import path.
    StateImportAbuse,
    /// No attack shape; trains the "not malicious" decision boundary.
    Benign,
}

impl FamilyId {
    /// Wire string (matches Python `FamilyId.value`).
    pub const fn as_str(self) -> &'static str {
        match self {
            FamilyId::ResponseAmp => "response_amp",
            FamilyId::ComputeAmp => "compute_amp",
            FamilyId::MemoryAmp => "memory_amp",
            FamilyId::ConnectionExhaustion => "connection_exhaustion",
            FamilyId::ConsensusAbuse => "consensus_abuse",
            FamilyId::GossipAbuse => "gossip_abuse",
            FamilyId::AuthBypass => "auth_bypass",
            FamilyId::RateLimiterBypass => "rate_limiter_bypass",
            FamilyId::ServiceMisconfig => "service_misconfig",
            FamilyId::Reconnaissance => "reconnaissance",
            FamilyId::SubscriptionCpuAmp => "subscription_cpu_amp",
            FamilyId::StateImportAbuse => "state_import_abuse",
            FamilyId::Benign => "benign",
        }
    }
}

/// One-line definition for each family. Mirrors
/// `bundle_spec.taxonomy.FAMILY_DEFINITIONS` in Python.
pub fn family_definitions() -> &'static [(FamilyId, &'static str)] {
    &[
        (
            FamilyId::ResponseAmp,
            "Small request → disproportionately large response bytes. \
             Attacker-side upstream cost is trivial; target-side egress \
             dominates.",
        ),
        (
            FamilyId::ComputeAmp,
            "Small request → disproportionate server-side CPU. Input size \
             is small, processing cost is large (filter misses, parser \
             pathologies, verification fallbacks, sync-over-async wedges).",
        ),
        (
            FamilyId::MemoryAmp,
            "Unbounded accumulation of state in process memory. The \
             attacker's input does not need to shrink, but every subsequent \
             observation adds to retained state without bound.",
        ),
        (
            FamilyId::ConnectionExhaustion,
            "Per-connection or per-subscription resource leak. Opening or \
             subscribing consumes quota that isn't released on disconnect.",
        ),
        (
            FamilyId::ConsensusAbuse,
            "Validator-authenticated misbehaviour inside the consensus \
             protocol — equivocations, backpressure injection, proposal \
             poisoning. Presupposes validator key material.",
        ),
        (
            FamilyId::GossipAbuse,
            "Unauthenticated p2p / gossip protocol abuse — state-sync \
             floods, request amplification via public peer endpoints.",
        ),
        (
            FamilyId::AuthBypass,
            "Missing or broken authentication on a control, admin, or \
             first-party API surface that should not be reachable without \
             credentials.",
        ),
        (
            FamilyId::RateLimiterBypass,
            "Defect in rate-limiting logic allowing amplified or higher-\
             than-intended traffic through the limiter.",
        ),
        (
            FamilyId::ServiceMisconfig,
            "Exploitable misconfiguration of an ancillary daemon that \
             validators commonly co-locate (Redis, Elasticsearch, Grafana, \
             Prometheus, SSH, …).",
        ),
        (
            FamilyId::Reconnaissance,
            "Pre-discovery and target-mapping activity against validator \
             infrastructure. Currently unpopulated in the v0.1.0 reference \
             corpus.",
        ),
        (
            FamilyId::SubscriptionCpuAmp,
            "A subscription or streaming filter whose per-notification \
             server-side cost is disproportionate to the trivial cost of \
             subscribing. A wide filter makes every matching event drive \
             expensive server work, so one cheap subscribe sustains \
             unbounded CPU.",
        ),
        (
            FamilyId::StateImportAbuse,
            "Abuse of a node's state / snapshot bootstrap-import path. A \
             malformed or oversized state artefact (snapshot, append-vec, \
             ledger segment) fed to the import / reconstruction pipeline \
             triggers a crash, panic, or resource blow-up during \
             deserialisation — before the artefact is fully validated.",
        ),
        (
            FamilyId::Benign,
            "No attack shape. Present to train the 'not malicious' decision \
             boundary.",
        ),
    ]
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn family_id_serialises_to_wire_string() {
        for (variant, expected) in [
            (FamilyId::ResponseAmp, "response_amp"),
            (FamilyId::ComputeAmp, "compute_amp"),
            (FamilyId::MemoryAmp, "memory_amp"),
            (FamilyId::ConnectionExhaustion, "connection_exhaustion"),
            (FamilyId::ConsensusAbuse, "consensus_abuse"),
            (FamilyId::GossipAbuse, "gossip_abuse"),
            (FamilyId::AuthBypass, "auth_bypass"),
            (FamilyId::RateLimiterBypass, "rate_limiter_bypass"),
            (FamilyId::ServiceMisconfig, "service_misconfig"),
            (FamilyId::Reconnaissance, "reconnaissance"),
            (FamilyId::SubscriptionCpuAmp, "subscription_cpu_amp"),
            (FamilyId::StateImportAbuse, "state_import_abuse"),
            (FamilyId::Benign, "benign"),
        ] {
            assert_eq!(serde_json::to_value(variant).unwrap(), json!(expected));
            assert_eq!(variant.as_str(), expected);
        }
    }

    #[test]
    fn family_id_deserialises_from_wire_string() {
        for s in [
            "response_amp",
            "compute_amp",
            "memory_amp",
            "connection_exhaustion",
            "consensus_abuse",
            "gossip_abuse",
            "auth_bypass",
            "rate_limiter_bypass",
            "service_misconfig",
            "reconnaissance",
            "benign",
        ] {
            let _: FamilyId = serde_json::from_str(&format!("\"{}\"", s)).unwrap();
        }
    }

    #[test]
    fn family_id_rejects_unknown_value() {
        let result: Result<FamilyId, _> = serde_json::from_str("\"not_a_family\"");
        assert!(result.is_err());
    }

    #[test]
    fn family_definitions_covers_all_variants() {
        let defs = family_definitions();
        assert_eq!(defs.len(), 13);
        for variant in [
            FamilyId::ResponseAmp,
            FamilyId::ComputeAmp,
            FamilyId::MemoryAmp,
            FamilyId::ConnectionExhaustion,
            FamilyId::ConsensusAbuse,
            FamilyId::GossipAbuse,
            FamilyId::AuthBypass,
            FamilyId::RateLimiterBypass,
            FamilyId::ServiceMisconfig,
            FamilyId::Reconnaissance,
            FamilyId::SubscriptionCpuAmp,
            FamilyId::StateImportAbuse,
            FamilyId::Benign,
        ] {
            assert!(
                defs.iter().any(|(v, _)| *v == variant),
                "missing family definition for {:?}",
                variant
            );
        }
    }
}
